### PR TITLE
Allow customization of created test cycle

### PR DESCRIPTION
**Pull request by johnathafelix**
I've added a few more fields in the post actions form that will allow users to customize the test cycle that is created.

This is how it looks like:
Screen Shot 2021-11-12 at 17 48 14

The new fields will be displayed only if the checkbox is checked:
Screen Shot 2021-11-12 at 17 50 01

There are placeholders for every field, so the user gets an example of the accepted input format.

As this is a feature only available for Zephyr Scale Cloud, setting those fields for a Server/DC instance won't do anything. It won't cause any errors, it will just be ignored.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -39,7 +39,7 @@
     </scm>
 
     <properties>
-        <jenkins.version>2.235.5</jenkins.version>
+        <jenkins.version>2.303.1</jenkins.version>
```
Comment (johnathafelix) [inline pom.xml:42]: Some Jenkins plugins were failing to start due to the version, that's why I upgraded it here.

File: src/main/java/com/adaptavist/tm4j/jenkins/utils/GsonUtils.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package com.adaptavist.tm4j.jenkins.utils;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializer;
+
+public class GsonUtils {
```
Comment (johnathafelix) [inline src/main/java/com/adaptavist/tm4j/jenkins/utils/GsonUtils.java:8]: We should probably make such conversion in the REST API/Backend. 

I think that it's reasonable to accept `50.0` as an integer value, for example.

Comment (nayeligarciac) [reply src/main/java/com/adaptavist/tm4j/jenkins/utils/GsonUtils.java:8]: @johnathafelix  how  is it going to differentiate between a Number or a Decimal Number Custom fields?

**Comment by nayeligarciac** [comment]
@johnathafelix  why didn’t you include any test in here ?

File: src/main/java/com/adaptavist/tm4j/jenkins/utils/GsonUtils.java (hunk 2 above)
Comment (chicofaccin) [reply src/main/java/com/adaptavist/tm4j/jenkins/utils/GsonUtils.java:8]: I agree. We can do this conversion in the API.

**Review by chicofaccin — COMMENTED** [review]
It think it would be good to add some unit tests

**Comment by chicofaccin** [comment]
> I've added a few more fields in the post actions form that will allow users to customize the test cycle that is created.
> 
> This is how it looks like: Screen Shot 2021-11-12 at 17 48 14
> 
> The new fields will be displayed only if the checkbox is checked: Screen Shot 2021-11-12 at 17 50 01
> 
> There are placeholders for every field, so the user gets an example of the accepted input format.
> 
> As this is a feature only available for Zephyr Scale Cloud, setting those fields for a Server/DC instance won't do anything. It won't cause any errors, it will just be ignored.

It would be great if you can add an example using pipelines https://github.com/jenkinsci/tm4j-automation-plugin#pipeline

File: pom.xml (hunk 1 above)
Comment (johnathafelix) [reply pom.xml:42]: @chicofaccin I think that would be okay but I followed the recommendation here: https://www.jenkins.io/doc/developer/plugin-development/choosing-jenkins-baseline

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/CustomTestCycle.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    public boolean isEmpty() {
+        return isBlank(name)
+            && isBlank(description)
+            && (customFields == null || customFields.size() == 0)
+            && (jiraProjectVersion == null || jiraProjectVersion == 0)
+            && (folderId == null || folderId == 0);
+    }
+
+    private Long convertToLongIfValid(final String value) {
+        if (isBlank(value)) {
```
[the remark this answers is not in the document]
Comment (johnathafelix) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/CustomTestCycle.java:62]: @chicofaccin not sure if we can... According to the documentation, `A decimal point is not a Unicode digit and returns false.`, so values such as `12.3` returns `false` there. See: https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#isNumeric-java.lang.CharSequence-

**Comment by johnathafelix** [comment]
> It would be great if you can add an example using pipelines https://github.com/jenkinsci/tm4j-automation-plugin#pipeline

@chicofaccin ah, I forgot about that. Thanks for pointing it out, good catch!

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/CustomTestCycle.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                    }.getType()
+                );
+
+            /*
+            For security reasons, Jenkins doesn't allow us to marshal
+            properties that are instances of any type of Gson classes.
+            That's why we need to convert Gson's LinkedTreeMap to a
+            HashMap. See https://jenkins.io/redirect/class-filter
+             */
+            return new HashMap<>(temp);
```
[the remark this answers is not in the document]
Comment (johnathafelix) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/CustomTestCycle.java]: Addressed!

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/CustomTestCycle.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (johnathafelix) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/CustomTestCycle.java:62]: But perhaps we don't even need to call `isBlank` here... I just checked and getting `Long.value()` of an empty string or null will throw `NumberFormatException`. So, I think we only need the `try/catch`.

File: pom.xml (hunk 1 above)
Comment (chicofaccin) [reply pom.xml:42]: @johnathafelix thanks. I didn't know about that. :)

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/CustomTestCycle.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (chicofaccin) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/CustomTestCycle.java:62]: You are right. 👍

File: src/main/java/com/adaptavist/tm4j/jenkins/utils/GsonUtils.java (hunk 2 above)
Comment (nayeligarciac) [reply src/main/java/com/adaptavist/tm4j/jenkins/utils/GsonUtils.java:8]: @johnathafelix  got it!  .. Sorry,   I thought that all the numbers to `50.0`

**Review by nayeligarciac — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -51,10 +51,17 @@ pipeline {
...
+            format: 'Cucumber', 
+            filePath: 'target/cucumber/*.json', 
+            autoCreateTestCases: false, 
+              customTestCycle: [
+                name: 'Jenkins Build',
+                description: 'Results from Jenkins Build', 
+                jiraProjectVersion: '10001', 
+                folderId: '3040527', 
+                customFields: '{"number":50,"single-choice":"option1","checkbox":true,"userpicker":"5f8b5cf2ddfdcb0b8d1028bb","single-line":"a text line","datepicker":"2020-01-25","decimal":10.55,"multi-choice":["choice1","choice3"],"multi-line":"first line<br />second line"}'
+              ]
```
Comment (chicofaccin) [inline README.md:64]: 👏

**Review by chicofaccin — APPROVED** [review]

**Comment by johnathafelix** [comment]
@nayeligarciac @chicofaccin this PR is already approved but just wanted to let you know that I have added unit tests to those classes 😉

**Pull request merged**